torch/lib/TensorLib/templates/StorageDerived.cpp

```diff
@@ -0,0 +1,91 @@
+#include "TensorLib/${Storage}.h"
+#include "TensorLib/HalfConvert.h"
+
+namespace tlib {
+
+${Storage}::${Storage}(Context* context):
+    storage(${THStorage}_new(${state})), context(context) {}
+
+${Storage}::${Storage}(Context* context, ${THStorage}* storage):
+    storage(storage), context(context) {}
+
+${Storage}::${Storage}(Context* context, std::size_t storage_size)
+  : storage(${THStorage}_newWithSize(${state,} storage_size)), context(context) {}
+
+${Storage}::~${Storage}() {
+  ${THStorage}_free(${state,} storage);
+}
+
+std::size_t ${Storage}::elementSize() const {
+  return sizeof(${ScalarType});
+}
+
+std::size_t ${Storage}::size() const {
+  return storage->size;
+}
+
+void* ${Storage}::data() {
+  return storage->data;
+}
+
+const void* ${Storage}::data() const {
+  return storage->data;
+}
+
+auto ${Storage}::retain() -> ${Storage}& {
+  ${THStorage}_retain(${state,} storage);
+  return *this;
+}
+
+auto ${Storage}::free() -> ${Storage}& {
+  ${THStorage}_free(${state,} storage);
+  return *this;
+}
+
+auto ${Storage}::resize(long new_size) -> ${Storage}& {
+  ${THStorage}_resize(${state,} storage, new_size);
+  return *this;
+}
+
+auto ${Storage}::fill(Scalar value) -> ${Storage}& {
+  ${THStorage}_fill(${state,} storage, ${to_th_half}(value.to${ScalarName}()));
+  return *this;
+}
+
+auto ${Storage}::set(std::size_t ind, Scalar value) -> ${Storage}& {
+  ${THStorage}_set(${state,} storage, ind, ${to_th_half}(value.to${ScalarName}()));
+  return *this;
+}
+
+auto ${Storage}::fast_set(std::size_t ind, Scalar value) -> ${Storage}& {
+  throw std::runtime_error("unsupported operation 'fast_set'");
```

Comment:
it has to behave differently than THPP because it will need to take a scalar_type template argument. We'll make sure to have it when we start replacing uses of THPP.